ml-algorithms/src/main/java/org/opensearch/ml/engine/agents/AgentContextUtil.java

```diff
@@ -0,0 +1,201 @@
+import org.opensearch.ml.common.conversation.Interaction;
+import org.opensearch.ml.common.hooks.EnhancedPostToolEvent;
+import org.opensearch.ml.common.hooks.HookRegistry;
+import org.opensearch.ml.common.hooks.PreLLMEvent;
+import org.opensearch.ml.common.spi.memory.Memory;
+import org.opensearch.ml.common.utils.StringUtils;
+import org.opensearch.ml.engine.memory.ConversationIndexMemory;
+
+public class AgentContextUtil {
+    private static final Logger log = LogManager.getLogger(AgentContextUtil.class);
+
+    public static ContextManagerContext buildContextManagerContextForToolOutput(
+        String toolOutput,
+        Map<String, String> parameters,
+        List<MLToolSpec> toolSpecs,
+        Memory memory
+    ) {
+        ContextManagerContext.ContextManagerContextBuilder builder = ContextManagerContext.builder();
+
+        String systemPrompt = parameters.get(SYSTEM_PROMPT_FIELD);
+        if (systemPrompt != null) {
+            builder.systemPrompt(systemPrompt);
+        }
+
+        String userPrompt = parameters.get(QUESTION);
+        if (userPrompt != null) {
+            builder.userPrompt(userPrompt);
+        }
+
+        if (toolSpecs != null) {
+            builder.toolConfigs(toolSpecs);
+        }
+
+        Map<String, String> contextParameters = new HashMap<>();
+        contextParameters.putAll(parameters);
+        contextParameters.put("_current_tool_output", toolOutput);
+        builder.parameters(contextParameters);
+
+        return builder.build();
+    }
+
+    public static Object extractProcessedToolOutput(ContextManagerContext context) {
+        if (context.getParameters() != null) {
+            return context.getParameters().get("_current_tool_output");
+        }
+        return null;
+    }
+
+    public static Object extractFromContext(ContextManagerContext context, String key) {
+        if (context.getParameters() != null) {
+            return context.getParameters().get(key);
+        }
+        return null;
+    }
+
+    public static ContextManagerContext buildContextManagerContext(
+        Map<String, String> parameters,
+        List<String> interactions,
+        List<MLToolSpec> toolSpecs,
+        Memory memory
+    ) {
+        ContextManagerContext.ContextManagerContextBuilder builder = ContextManagerContext.builder();
+
+        String systemPrompt = parameters.get(SYSTEM_PROMPT_FIELD);
+        if (systemPrompt != null) {
+            builder.systemPrompt(systemPrompt);
+        }
+
+        String userPrompt = parameters.get(QUESTION);
+        if (userPrompt != null) {
+            builder.userPrompt(userPrompt);
+        }
+
+        if (memory instanceof ConversationIndexMemory) {
+            String chatHistory = parameters.get(CHAT_HISTORY);
+            if (chatHistory != null) {
+                List<Interaction> chatHistoryList = new ArrayList<>();
+                builder.chatHistory(chatHistoryList);
+            }
+        }
```

Comment:
Is this a bug? Can you double check. chatHistoryList is empty and never adds anything. Are you trying to do the following: 
```
String chatHistoryJson = parameters.get(CHAT_HISTORY);
if (chatHistoryJson != null) {
    List<Interaction> chatHistory = parseChatHistory(chatHistoryJson);
    if (!chatHistory.isEmpty()) {
        builder.chatHistory(chatHistory);
    }
}
```